### PR TITLE
State changes outport

### PR DESCRIPTION
**Pull request by cristure**
## Reasoning behind the pull request
- State changes will be also exported through the host driver (and possible other methods)
  
## Proposed changes
- Add a new method to the `Driver` interface that will send the state changes collected.

## Testing procedure
- Will need to run a local-testnet along with an observer node that will have the configuration for exporting through host driver or other means.

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

**Comment by cristure** [comment]
There are somethings that need to be addressed. Mainly about coupling `stateChangesCollectore` in the `outportDriver`. The other way around is not possible (there is a [TODO](https://github.com/multiversx/mx-chain-go/blob/03102683d950a530a0faea4e0c75cf3a88363f54/state/stateChanges/writeCollector.go#L75) left, implying the outportDriver should be injected as a dependency in the `stateChangesCollector`, but doing this will result in a import cycle and therefore the code won't compile. Therefore, I injected the `stateChangesCollector` in the `outportDriver`, but that will require some additional changes on the integration between these two.

**Comment by cristure** [comment]
Also moved some structs in proto definition files. PR can be seen [here](https://github.com/multiversx/mx-chain-core-go/pull/320).

Moreover, the mx-chain-go-storage had to be updated with the subsequent change in the proto strucutres. PR [here](https://github.com/multiversx/mx-chain-storage-go/pull/53)

File: scripts/testnet/variables.sh [hunk 1]
```diff
@@ -12,7 +12,7 @@ export USE_PROXY=1
 
 # Enable the MultiversX Transaction Generator. Note that this is a private
 # repository (mx-chain-txgen-go).
-export USE_TXGEN=0
+export USE_TXGEN=1
```
Comment (miiu96) [inline scripts/testnet/variables.sh:15]: why did you change these values?

File: state/disabled/disabledStateChangesCollector.go [hunk 2]
```diff
@@ -47,6 +49,10 @@ func (d *disabledStateChangesCollector) Publish() error {
 	return nil
 }
 
+func (d *disabledStateChangesCollector) GetStateChangesForTxs() map[string]*stateChange.StateChanges {
```
Comment (miiu96) [inline state/disabled/disabledStateChangesCollector.go:52]: missing comment

File: state/stateChanges/dataAnalysisCollector.go [hunk 3]
```diff
@@ -186,6 +187,10 @@ func (scc *dataAnalysisCollector) Publish() error {
 	return nil
 }
 
+func (scc *dataAnalysisCollector) GetStateChangesForTx() []StateChange {
```
Comment (miiu96) [inline state/stateChanges/dataAnalysisCollector.go:190]: missing comment

File: state/stateChanges/writeCollector.go [hunk 4]
```diff
@@ -136,6 +103,73 @@ func (scc *stateChangesCollector) getStateChangesForTxs() ([]StateChangesForTx,
 	return stateChangesForTxs, nil
 }
 
+func (scc *stateChangesCollector) GetStateChangesForTxs() map[string]*data.StateChanges {
```
Comment (miiu96) [inline state/stateChanges/writeCollector.go:106]: missing comment

File: state/stateChanges/writeCollector.go [hunk 5]
```diff
@@ -136,6 +103,73 @@ func (scc *stateChangesCollector) getStateChangesForTxs() ([]StateChangesForTx,
...
+					MainTrieKey:     stateChange.GetMainTrieKey(),
+					MainTrieVal:     stateChange.GetMainTrieVal(),
+					Operation:       stateChange.GetOperation(),
+					DataTrieChanges: stateChange.GetDataTrieChanges(),
+				},
+			)
+		}
+	}
+
+	//for i := 0; i < len(scc.stateChanges); i++ {
```
Comment (miiu96) [inline state/stateChanges/writeCollector.go:143]: what is this commented code ?

File: outport/factory/outportFactory.go [hunk 6]
```diff
@@ -73,12 +74,14 @@ func createAndSubscribeElasticDriverIfNeeded(
...
-	elasticDriver, err := indexerFactory.NewIndexer(args)
+	//TODO: this needs discussed.
+	_, err := indexerFactory.NewIndexer(args)
 	if err != nil {
 		return err
 	}
 
-	return outport.SubscribeDriver(elasticDriver)
+	return nil
+	//return outport.SubscribeDriver(elasticDriver)
```
[the remark this answers is not in the document]
Comment (cristure) [reply outport/factory/outportFactory.go]: reverted. sorry for that.

File: outport/factory/hostDriverFactory.go [hunk 7]
```diff
@@ -4,15 +4,19 @@ import (
...
+
 	logger "github.com/multiversx/mx-chain-logger-go"
 )
 
 type ArgsHostDriverFactory struct {
-	HostConfig config.HostDriversConfig
-	Marshaller marshal.Marshalizer
+	HostConfig            config.HostDriversConfig
+	Marshaller            marshal.Marshalizer
+	StateChangesCollector state.StateChangesCollector
```
[the remark this answers is not in the document]
Comment (cristure) [reply outport/factory/hostDriverFactory.go]: nope, removed it altoghether

**Review by miiu96 — APPROVED** [review]

File: outport/process/outportDataProvider.go [hunk 8]
```diff
@@ -134,6 +139,9 @@ func (odp *outportDataProvider) PrepareOutportSaveBlockData(arg ArgPrepareOutpor
 		return nil, err
 	}
 
+	stateChanges := odp.stateChangesCollector.GetStateChangesForTxs()
+	odp.stateChangesCollector.Reset()
```
Comment (AdoAdoAdo) [inline outport/process/outportDataProvider.go:143]: I would move the reset on the commit block, after the indexing part. 
Otherwise, in case there is no client for the outport block, you would not end up in this method and the reset would not be done.

File: state/accountsDB.go [hunk 9]
```diff
@@ -297,12 +298,12 @@ func (adb *AccountsDB) SaveAccount(account vmcommon.AccountHandler) error {
...
 }
 
-func (adb *AccountsDB) saveCodeAndDataTrie(oldAcc, newAcc vmcommon.AccountHandler) ([]stateChanges.DataTrieChange, error) {
+func (adb *AccountsDB) saveCodeAndDataTrie(oldAcc, newAcc vmcommon.AccountHandler) ([]*stateChange.DataTrieChange, error) {
 	baseNewAcc, newAccOk := newAcc.(baseAccountHandler)
 	baseOldAccount, _ := oldAcc.(baseAccountHandler)
 
 	if !newAccOk {
-		return nil, nil
+		return make([]*stateChange.DataTrieChange, 0), nil
```
Comment (AdoAdoAdo) [inline state/accountsDB.go:306]: Is this required because of a panic?

File: state/accountsDB.go [hunk 10]
```diff
@@ -918,7 +918,8 @@ func (adb *AccountsDB) commit() ([]byte, error) {
 		return nil, err
 	}
 
-	adb.stateChangesCollector.Reset()
+	//TODO: discuss the workflow. If reset here, the outport driver won't be able to pick up the changes.
+	//adb.stateChangesCollector.Reset()
```
Comment (AdoAdoAdo) [inline state/accountsDB.go:922]: I would move the reset in the processor (shardProcessor & metaProcessor) in the CommitBlock, after calling `sp.indexBlockIfNeeded(bodyHandler, headerHash, headerHandler, lastBlockHeader)` on shardProcessor, and  
after calling `mp.indexBlock(header, headerHash, body, lastMetaBlock, notarizedHeadersHashes, rewardsTxs)` on metaProcessor

File: state/stateChanges/writeCollector.go [hunk 11]
```diff
@@ -136,6 +103,47 @@ func (scc *stateChangesCollector) getStateChangesForTxs() ([]StateChangesForTx,
 	return stateChangesForTxs, nil
 }
 
+// GetStateChangesForTxs will retrieve the state changes linked with the tx hash.
+func (scc *stateChangesCollector) GetStateChangesForTxs() map[string]*data.StateChanges {
+	scc.stateChangesMut.Lock()
```
Comment (AdoAdoAdo) [inline state/stateChanges/writeCollector.go:108]: I think a RLock would be sufficient, as you are just reading the state changes not modifying

Comment (cristure) [reply state/stateChanges/writeCollector.go:108]: we are actually appending to the map in case the hash is already present. Can be seen at state/stateChanges/writeCollector.go:130

File: state/accountsDB.go (hunk 10 above)
Comment (cristure) [reply state/accountsDB.go:922]: moved

File: state/accountsDB.go (hunk 9 above)
Comment (cristure) [reply state/accountsDB.go]: `TestMetaProcessor_CommitBlockOkValsShouldWork` yes, this test would fail with the previous implementation.

Comment (AdoAdoAdo) [reply state/accountsDB.go]: @BeniaminDrasovean can you check if this change would be problematic?
do we need to change the test instead?

File: state/accountsDB.go (hunk 10 above)
Comment (AdoAdoAdo) [reply state/accountsDB.go:922]: don't forget to remove also the commented code after we agree on the workflow.

File: state/stateChanges/writeCollector.go [hunk 12]
```diff
@@ -136,6 +103,47 @@ func (scc *stateChangesCollector) getStateChangesForTxs() ([]StateChangesForTx,
...
+					TxHash:          stateChange.GetTxHash(),
+					MainTrieKey:     stateChange.GetMainTrieKey(),
+					MainTrieVal:     stateChange.GetMainTrieVal(),
+					Operation:       stateChange.GetOperation(),
+					DataTrieChanges: stateChange.GetDataTrieChanges(),
+				},
+			},
+			}
+		} else {
+			stateChangesForTxs[txHash].StateChanges = append(sc.StateChanges,
```
Comment (AdoAdoAdo) [inline state/stateChanges/writeCollector.go:130]: here we should not append to the sc.StateChanges, but to the stateChangesForTxs[txHash].StateChanges
then we can have the RWMutex as well.

Comment (AdoAdoAdo) [reply state/stateChanges/writeCollector.go:130]: can you also add some unit tests for GetStateChangesForTxs ?

File: outport/host/errors.go [hunk 13]
```diff
@@ -7,3 +7,6 @@ var ErrHostIsClosed = errors.New("server is closed")
 
 // ErrNilHost signals that a nil host has been provided
 var ErrNilHost = errors.New("nil host provided")
+
+// ErrNilStateChangesCollector signals that a nil state change collector has been provided.
+var ErrNilStateChangesCollector = errors.New("nil state changes collector provided")
```
Comment (BeniaminDrasovean) [inline outport/host/errors.go:12]: This is not used anywhere.

File: state/accountsDB.go (hunk 9 above)
Comment (BeniaminDrasovean) [reply state/accountsDB.go]: It's ok both ways.

File: testscommon/state/stateChangesCollectorStub.go [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+import (
+	"github.com/multiversx/mx-chain-core-go/data/stateChange"
+	"github.com/multiversx/mx-chain-core-go/data/transaction"
+	vmcommon "github.com/multiversx/mx-chain-vm-common-go"
+
+	"github.com/multiversx/mx-chain-go/state/stateChanges"
+)
+
+type StateChangesCollectorStub struct {
```
Comment (BeniaminDrasovean) [inline testscommon/state/stateChangesCollectorStub.go:11]: Missing comment here.

File: state/stateChanges/writeCollector.go (hunk 12 above)
Comment (cristure) [reply state/stateChanges/writeCollector.go:130]: oh, you are right. Sorry for that, adding unit test as well.

Comment (cristure) [reply state/stateChanges/writeCollector.go:130]: but it is very confusing as it is. Will use he same variable name for both.

File: outport/host/errors.go (hunk 13 above)
Comment (cristure) [reply outport/host/errors.go:12]: it is used here: state/accountsDB.go:163

File: testscommon/state/stateChangesCollectorStub.go (hunk 14 above)
Comment (cristure) [reply testscommon/state/stateChangesCollectorStub.go:11]: added

File: state/accountsDB.go [hunk 15]
```diff
@@ -159,9 +160,7 @@ func checkArgsAccountsDB(args ArgsAccountsDB) error {
 	if check.IfNil(args.SnapshotsManager) {
 		return ErrNilSnapshotsManager
 	}
-	if check.IfNil(args.StateChangesCollector) {
-		return ErrNilStateChangesCollector
-	}
+	if check.IfNil(args.StateChangesCollector) { return ErrNilStateChangesCollector }
```
[the remark this answers is not in the document]
Comment (cristure) [reply state/accountsDB.go:163]: sure

File: process/block/metablock_test.go [hunk 16]
```diff
@@ -919,6 +920,9 @@ func TestMetaProcessor_CommitBlockStorageFailsForHeaderShouldNotReturnError(t *t
 		return &block.Header{}, []byte("hash"), nil
 	}
 	arguments.BlockTracker = blockTrackerMock
+	arguments.StateChangesCollector = &stateMock.StateChangesCollectorStub{
+		ResetCalled: func() {},
```
[the remark this answers is not in the document]
Comment (cristure) [reply process/block/metablock_test.go]: done

**Review by AdoAdoAdo — APPROVED** [review]
